tedana/utils/masking.py

```diff
@@ -0,0 +1,162 @@
+"""
+"""
+import numpy as np
+import pandas as pd
+import nibabel as nib
+
+
+def row_idx(arr1, arr2):
+    """
+    Get a 1D index of rows in arr1 that exist in arr2.
+
+    Parameters
+    ----------
+    arr1 : (X x 3) :obj:`numpy.ndarray`
+    arr2 : (Z x 3) :obj:`numpy.ndarray`
+
+    Returns
+    -------
+    idx : 1D :obj:`numpy.ndarray`
+        Index of rows in arr1 that exist in arr2.
+
+    Notes
+    -----
+    This works amazingly well, but is quite slow.
+    """
+    df1 = pd.DataFrame(arr1, columns=['x', 'y', 'z'], dtype=str)
+    df2 = pd.DataFrame(arr2, columns=['x', 'y', 'z'], dtype=str)
+
+    df1['unique_value'] = df1[['x', 'y', 'z']].apply(lambda x: '_'.join(x),
+                                                     axis=1)
+    df1 = df1[['unique_value']]
+    df1['idx1'] = df1.index
+    df1 = df1.set_index('unique_value')
+
+    df2['unique_value'] = df2[['x', 'y', 'z']].apply(lambda x: '_'.join(x),
+                                                     axis=1)
+    df2 = df2[['unique_value']]
+    df2['idx2'] = df2.index
+    df2 = df2.set_index('unique_value')
+
+    catted = pd.concat((df1, df2), axis=1, ignore_index=False)
+    if any(pd.isnull(catted['idx1'])):
+        raise Exception('We have a weird error where there is >=1 '
+                        'voxel in echo-specific mask outside of union mask.')
+
+    catted = catted.dropna()
+    catted = catted.sort_values(by=['idx2'])
+    rel_echo_idx = catted['idx1'].values
+    return rel_echo_idx
+
+
+def apply_mask_me(img, mask_img):
+    """
+    Apply multi-echo mask to set of nifti images. Mask may vary by echo.
+
+    Parameters
+    ----------
+    img : (X x Y x Z x E [x T]) :obj:`nibabel.nifti.Nifti1Image`
+        Data img.
+    mask_img : (X x Y x Z [x E]) :obj:`nibabel.nifti.Nifti1Image`
+        Mask img.
+
+    Returns
+    -------
+    masked_arr : (M x E [x T]) :obj:`numpy.ndarray`
+        Masked data. M refers to the number of voxels in the mask img.
+    """
+    if not isinstance(mask_img, nib.Nifti1Image):
+        raise TypeError('Provided mask is not a Nifti1Image.')
+    elif len(mask_img.shape) not in (3, 4):
+        raise ValueError('Mask must be 3D (X x Y x Z) or 4D (X x Y x Z x E).')
+
+    if not np.array_equal(img.affine, mask_img.affine):
+        raise ValueError('Input img affine must match mask_img affine.')
+
+    data = img.get_data()
+    if len(img.shape) == 4:
+        _, _, _, n_echos = img.shape
+        n_vols = 1
+        data = data[:, :, :, :, None]
+    elif len(img.shape) == 5:
+        _, _, _, n_echos, n_vols = img.shape
+    else:
+        raise ValueError('Input img must be 4D (X x Y x Z x E) or '
```

Comment:
Is this a valid error to pass back up to the user ? They'll have either passed in a zcat (` X x Y x Z x E x T`) or individual echo files, so I don't think they would interact with a `(X x Y x Z x E)` image..?

Reply:
`load_data2` would load zcat (`X x Y x Z*E x T`), the preferred 5D (`X x Y x Z x E x T`) and individual echo files all into the same `X x Y x Z x E x T`-ordered image type, so masking and unmasking would respectively take in and return only that image type. We could allow other image types, but I think that it's unnecessary. Multiple individual echo images could just be masked/unmasked with nilearn's functions, and we should have to deal with zcat images (even when we do have zcat files) because we'll have `load_data2`

Reply:
Oh sorry, I misinterpreted your point. I guess it would be rare to feed in an image that isn't `X x Y x Z x E x T` or `X x Y x Z x E`, but I think that just makes this error unlikely (but not invalid).

Reply:
Sorry for being unclear ! To clarify: shouldn't it be a `X x Y x Z x T` file, not a `X x Y x Z x E` file ? The former would be an individual echo file, the latter is something I don't think users will interact with.

Reply:
Completely agree, but I'm not sure how this error would make sense to a user, since they'll have only supplied the individual echo files or the zcat image (as far as I understand) ! The `X x Y x Z x E` file would then be created "under-the-hood" in a workflow.